src/ngraph_encapsulate_op.cc

```diff
@@ -165,51 +185,82 @@ class NGraphEncapsulateOp : public tf::OpKernel {
-      if (m_freshness_tracker->IsFresh(src_ptr, ng_function) &&
-          src_ptr == last_used_src_ptrs[i]) {
-        t->set_stale(false);
+      // At the first call of the ng_function, both last_src_ptr and
+      // last_tv shall point to null. Otherwise, they are retrived
+      // from cache.
+      void* last_src_ptr = input_caches[i].first;
+      std::shared_ptr<ng::runtime::TensorView> last_tv = input_caches[i].second;
+
+      void* current_src_ptr = (void*)tf::DMAHelper::base(&ctx->input(i));
+      std::shared_ptr<ng::runtime::TensorView> current_tv;
+
+      if (m_ng_backend_name == "CPU") {
+        // We need to check last_tv != nullptr, since there are cases where at
+        // the first call to the ng_function, both the current_src_ptr (when the
+        // input is a 0-sized tensor) and last_src_ptr (uninitialized at the
+        // first call) are nullptr
+        if (current_src_ptr == last_src_ptr && last_tv != nullptr) {
+          // Mark each tensor as non-stale if:
+          //   1. the freshness tracker says the tensor has not changed since
+          //      the last time ng_function was called, and
+          //   2. we are using the same tensor in this argument position as
+          //      the one we used last time ng_function was called.
+          if (m_freshness_tracker->IsFresh(current_src_ptr, ng_function)) {
+            last_tv->set_stale(false);
+          } else {
+            last_tv->set_stale(true);
+          }
+          current_tv = last_tv;
+        } else {
+          current_tv = m_ng_backend->create_tensor(ng_element_type, ng_shape,
+                                                   current_src_ptr);
+          current_tv->set_stale(true);
+        }
       } else {
-        t->set_stale(true);
-      }
-      last_used_src_ptrs[i] = src_ptr;
-      ng_inputs.push_back(t);
-    }
+        if (last_tv != nullptr) {
+          current_tv = last_tv;
+        } else {
+          current_tv = m_ng_backend->create_tensor(ng_element_type, ng_shape);
+        }
+        current_tv->write(current_src_ptr, 0, current_tv->get_element_count() *
+                                                  ng_element_type.size());
+      }  // if (m_ng_backend_name == "CPU")
+
+      input_caches[i] = std::make_pair(current_src_ptr, current_tv);
+      ng_inputs.push_back(current_tv);
+    }  // for (int i = 0; i < input_shapes.size(); i++)
 
     NGRAPH_VLOG(4) << "NGraphEncapsulateOp::Compute allocated argument tensors "
                       "for cluster "
                    << m_ngraph_cluster;
 
     // Allocate tensors for the results.
-    vector<shared_ptr<ng::runtime::TensorView>> outputs;
+    vector<shared_ptr<ng::runtime::TensorView>> ng_outputs;
+
+    std::vector<std::pair<void*, std::shared_ptr<ng::runtime::TensorView>>>&
+        output_caches = m_ng_function_output_cache_map[ng_function];
+    output_caches.resize(ng_function->get_output_size());
+
```

Comment:
We should be able to check if it's cached and goes to copy directly for GPU and NNP if it is.